client/context/query.go

```diff
@@ -72,22 +74,64 @@ func (ctx CLIContext) GetAccount(address []byte) (auth.Account, error) {
 }
 
 // GetFromAddress returns the from address from the context's name.
-func (ctx CLIContext) GetFromAddress() (from sdk.AccAddress, err error) {
-	if ctx.FromAddressName == "" {
-		return nil, errors.Errorf("must provide a from address name")
+func (ctx CLIContext) GetFromAddress() (sdk.AccAddress, error) {
+	if ctx.fromAddress != nil {
+		return ctx.fromAddress, nil
 	}
 
-	keybase, err := keys.GetKeyBase()
+	err := ctx.populateFromFields()
 	if err != nil {
 		return nil, err
 	}
 
-	info, err := keybase.Get(ctx.FromAddressName)
+	return ctx.fromAddress, nil
+}
+
+func (ctx CLIContext) GetFromName() (string, error) {
+	if ctx.fromName != "" {
+		return ctx.fromName, nil
+	}
+
+	err := ctx.populateFromFields()
 	if err != nil {
-		return nil, errors.Errorf("no key for: %s", ctx.FromAddressName)
+		return "", err
+	}
+
+	return ctx.fromName, nil
+}
+
+func (ctx CLIContext) populateFromFields() error {
+	if ctx.From == "" {
+		return errors.Errorf("must provide a from address or name")
 	}
 
-	return sdk.AccAddress(info.GetPubKey().Address()), nil
+	keybase, err := keys.GetKeyBase()
+	if err != nil {
+		return err
+	}
+
+	var info cskeys.Info
+
+	if strings.HasPrefix(ctx.From, sdk.Bech32PrefixAccAddr) {
```

Comment:
There's a conflict if the user enters a name that happens to be a Bech32 address, or one that starts with the Bech32PrefixAccAddr.  A few thoughts:

* Checking for the Bech32PrefixAccAddr prefix is not enough, we should also check for the "1" that follows, as per the bech32 spec.
* Maybe we should just try parsing the address via sdk.AccAddressFromBech32 and assume it's an address if there's no error.  (there's a checksum in bech32 so this should be enough).
* And, maybe we should instead check on the add/new side and ensure that you can't enter a valid Bech32 address as a name.